mergin/client_pull.py

```diff
@@ -0,0 +1,489 @@
+                        continue
+                    if diff == v['diff']['path']:
+                        diff_file['version'] = k
+                        diff_file['diff'] = v['diff']
+                fetch_files.append(diff_file)
+        else:
+            f['version'] = server_version
+            fetch_files.append(f)
+
+    files_to_merge = []   # list of FileToMerge instances
+
+    for file in fetch_files:
+        diff_only = 'diffs' in file
+        items = _download_items(file, temp_dir, diff_only)
+        
+        # figure out destination path for the file
+        file_dir = os.path.dirname(os.path.normpath(os.path.join(temp_dir, file['path'])))
+        basename = os.path.basename(file['diff']['path']) if diff_only else os.path.basename(file['path'])
+        dest_file_path = os.path.join(file_dir, basename)
+        files_to_merge.append( FileToMerge(dest_file_path, items) )
+
+
+    # make sure we can update geodiff reference files (aka. basefiles) with diffs or
+    # download their full versions so we have them up-to-date for applying changes
+    basefiles_to_patch = []  # list of tuples (relative path within project, list of diff files in temp dir to apply)
+    for file in pull_changes['updated']:
+        if 'diffs' not in file:
+            continue  # this is only for diffable files (e.g. geopackages)
+
+        basefile = mp.fpath_meta(file['path'])
+        if not os.path.exists(basefile):
+            # The basefile does not exist for some reason. This should not happen normally (maybe user removed the file
+            # or we removed it within previous pull because we failed to apply patch the older version for some reason).
+            # But it's not a problem - we will download the newest version and we're sorted.
+            items = _download_items(file, temp_dir, diff_only=False)
+            dest_file_path = mp.fpath(file["path"], temp_dir)
+            #dest_file_path = os.path.join(os.path.dirname(os.path.normpath(os.path.join(temp_dir, file['path']))), os.path.basename(file['path']))
+            files_to_merge.append( FileToMerge(dest_file_path, items) )
+            continue   
+
+        basefiles_to_patch.append( (file['path'], file['diffs']) )
+
+    # make a single list of items to download
+    total_size = 0
+    download_list = []
+    for file_to_merge in files_to_merge:
+        download_list.extend(file_to_merge.downloaded_items)
+        for item in file_to_merge.downloaded_items:
+            total_size += item.size
+
+    job = PullJob(project_path, pull_changes, total_size, server_version, files_to_merge, download_list, temp_dir, mp, server_info, basefiles_to_patch)
+
+    # start download
+    job.executor = concurrent.futures.ThreadPoolExecutor(max_workers=4)
+    job.futures = []
+    for item in download_list:
+        future = job.executor.submit(_do_download, item, mc, project_path, job)
+        job.futures.append(future)
+    
+    return job
+
+
+def pull_project_wait(job):
```

Comment:
these wait, is_running and cancel functions are pretty much the same for download, pull and push, maybe we could use some generic ones?

Reply:
They may become more complex in the future (e.g. if we make also "project info" request asynchronous) and thus I would prefer to keep them separate...